src/common/framework/gpu/device_pool.ts

```diff
@@ -79,14 +79,21 @@ export class DevicePool {
   // Gets a device and creates a DeviceHolder.
   // If the device is lost, DeviceHolder.lostReason gets set.
   private static async makeHolder(): Promise<DeviceHolder> {
-    const gpu = getGPU();
-    const adapter = await gpu.requestAdapter();
+    const gpu = await getGPU();
+    const window = new WebGPUWindow({
+      width: 640,
+      height: 480,
+      title: 'WebGPU',
+    });
```

Comment:
Yes, due to the native heritage of the WebGPU NAPI addon, a graphics context (window) cannot be expected to pre-exist like with in-browser implmentations of WebGPU.

Ideally we should like to be able to run headless for Compute API applications of WebGPU.

Reply:
We definitely want it to be possible to use WebGPU in native without a window, and we should be able to use everything except Surface/SwapChain. I'm not sure where Dawn is wrt that goal though. (@austinEng?)

Reply:
Sorry, not clear what the question is. Dawn does work headless without a window exactly as Kai said. Most of our tests run this way.

Reply:
Thought we opened a window during test startup (at least on linux, one for Vulkan and one for OpenGL). Are those only used for tests that test window stuff? Or do we not have them anymore?

Reply:
We fixed this at some point, so now only OpenGL does it. We should only need a window specifically for tests that use a wgpu::Surface. Tests even work if there's no GPU using SwiftShader!

Reply:
Great, sounds like that won't be a problem when we get to fixing this.

Reply:
Where does CTS call device.tick ?
I can't find it in the code base. It seems to be required for native implementations.
The devices in the CTS pool should be ticked on process.nextTick, I think.

Reply:
It never calls device.tick because it doesn't exist in the JS API. Kai mentioned somewhere NAPI could call it at some interval, perhaps using process.nextTick as you suggested.

Reply:
Thread ended up getting split. Here was my comment:

---

It doesn't; there's no tick() in the JS API. Instead the implementation is supposed to figure out when to tick. In Chrome we currently do this dumbly, on a periodic basis, which has more overhead than we'd like. For NAPI I guess there are two options to start with:
- Expose it to JS and call it with setInterval or something
- Set up a setInterval-like timer in C++ during creation of the device

Reply:
Something invisible to JS would be necessary in the long run; perhaps the second option there is good enough.

Reply:
It is exposed in the WebGPU-NAPI addon. Our native evaluator simply traverses a primitive device pool and calls it for each device. 

I think we are closing on this. Will try it out after the meeting today.

Reply:
Something I wanted to clarify, what does the pool in your evaluator do? In our case the pool will:
- Hold multiple devices so multiple tests can run at once (or a test can use multiple devices)
- Reduce the number of devices created
- Possibly eventually, cache device creations with different options (combinations of adapter, limits, and extensions) (which would really be one pool for each combination of options)

Is your pool for using multiple adapters? Or some other purpose?

Reply:
I am having an issue with CTS - does it override the event loop?

- setTimeout(fn, 0) is banned 
- setInterval(fn, 0) does not fire callback
- process.nextTick(fn) does not fire callback

In our evaluator, we have a very primitive device pool than simply adds and removes devices to an array and then we recursively call process.nextTick (in Node) or setTimeout(fn, 0) in the native evaluator.

Reply:
No, we don't override the event loop. setTimeout is banned because WPT (web-platform-tests) has a lint against using it, but it's fine to use outside of tests and you can override it somehow.

setInterval should work. I tried a quick test of:
`setInterval(() => { console.log('a'); }, 500);`
at the bottom of `cmdline.ts`, and it worked fine. Do you have synchronous code running that's blocking the event loop?

---

I guess my bigger question is, why does your evaluator need multiple devices?

Reply:
Could you please try to run setInterval within device_pool.ts? That is where I am having the issue.

The last call to our code before setInterval is requestDevice, which you can see [here](https://github.com/gpuweb/cts/blob/3a45e00eed389d244a994ebc024f3032d88a1042/src/common/framework/gpu/implementation.ts#L56) returns a Promise, so I don't think that is the issue, again this works fine in our evaluator natively and in node.
 
Our evaluator is designed for the [DCP API (Distributed Compute Protocol)](https://docs.dcp.dev/index.html)

Our goal is to be able to run WebGPU Compute API pipelines on any and all GPUs available on a given host. This is why we are pushing for multi-GPU access in WebGPU (and have a primitive device pool designed for that).

I will try to move the timeout to implemtentation.ts (I can attach it to a device instance to see if that gets us forward).

Reply:
> Could you please try to run setInterval within device_pool.ts? That is where I am having the issue.

Sorry, I can't, because I still get a crash when I try to run this.

Reply:
I can run `webgpu:examples:basic:*` with a setInterval in DevicePool.acquire and that does work as expected.